### PR TITLE
cluster-init: generate Certificates

### PR DESCRIPTION
**Pull request by danilo-gemoli**
Generate `certificates.cert-manager.io` manifests into `assets/cert-manager` directory.
`cluster-init` extracts the following information at runtime:
- DNS base domain: get from the `baseDomain` stanza of `cluster-install.yaml` stored into `oc -n kube-system get cm/cluster-config-v1` 
- Image Registry public host: get from the `publicDockerImageRepository` stanza of any IS from the `openshift` namespace. I didn't come up with this idea, I've stolen from [here](https://github.com/openshift/oc/blob/e75d922ae2a2af65a647dd6c094be9e78d49f7d4/pkg/cli/registry/info/info.go#L1), thanks `oc` guys!

[Check this out](https://github.com/openshift/release/blob/b9355499f2cbcba18f7d90b9bca44bbc0cbbbe59/clusters/build-clusters/build11/cert-manager/certificate.yaml) as an example.

### CODE REVIEW

File: pkg/clustermgmt/onboard/certificate.go [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+	installConfig := struct {
+		BaseDomain string `json:"baseDomain"`
+	}{}
+	if err := yaml.Unmarshal([]byte(installConfigRaw), &installConfig); err != nil {
+		return "", fmt.Errorf("unmarshall install config: %w", err)
+	}
+	return installConfig.BaseDomain, nil
+}
+
+func (s *certificateStep) imageRegistryPublicHost(ctx context.Context, client ctrlruntimeclient.Client) (string, error) {
```
Comment (jmguzik) [inline pkg/clustermgmt/onboard/certificate.go:94]: locateImage... ? I had trouble to understand what this does

File: pkg/clustermgmt/onboard/certificate.go [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+	outputPath := CertificateManifestPath(s.clusterInstall.Onboard.ReleaseRepo, s.clusterInstall.ClusterName)
+	if err := s.writeManifest(outputPath, manifestMarshaled, 0644); err != nil {
+		return fmt.Errorf("write template %s: %w", outputPath, err)
+	}
+
+	log.WithField("certificate", outputPath).Info("certificates generated")
+	return nil
+}
+
+func (s *certificateStep) baseDomain(ctx context.Context, client ctrlruntimeclient.Client) (string, error) {
```
Comment (jmguzik) [inline pkg/clustermgmt/onboard/certificate.go:68]: establishBaseDomain?

File: pkg/clustermgmt/onboard/certificate.go [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+	if s.clusterInstall.Onboard.Certificate.BaseDomains != nil {
+		if domain, ok := s.clusterInstall.Onboard.Certificate.BaseDomains[s.clusterInstall.ClusterName]; ok {
+			s.log.Info("override base domain from config")
```
Comment (jmguzik) [inline pkg/clustermgmt/onboard/certificate.go:71]: log could be more descriptive with values

**Review by jmguzik — APPROVED** [review]
Some comments about clarity but generally LGTM. Can be fixed in next PR

File: pkg/clustermgmt/onboard/certificate.go [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+	if !(*s.clusterInstall.Onboard.OSD || *s.clusterInstall.Onboard.Hosted || *s.clusterInstall.Onboard.Unmanaged) {
+		manifests = append(manifests, apiServerCert, appsCert)
+	}
+	manifests = append(manifests, imageRegistryCert)
+
+	return manifests
+}
+
+func (s *certificateStep) clusterIssuerOrDefault(certificate, def string) string {
```
[the remark this answers is not in the document]
Comment (danilo-gemoli) [reply pkg/clustermgmt/onboard/certificate.go:200]: I try to follow this [style guide](https://google.github.io/styleguide/go/decisions#getters), `get...` is not recommended in Go

File: pkg/clustermgmt/onboard/certificate.go [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+	ci := s.clusterInstall.Onboard.Certificate.ClusterIssuer
+	if cluster, ok := ci[s.clusterInstall.ClusterName]; ok {
+		if clusterIssuer, ok := cluster[certificate]; ok {
+			return clusterIssuer
+		}
+	}
+	return def
+}
+
+func (s *certificateStep) projectLabelOrDefault(certificate, defKey, defValue string) (string, string) {
```
[the remark this answers is not in the document]
Comment (danilo-gemoli) [reply pkg/clustermgmt/onboard/certificate.go:210]: See answer above ^

**Pull request merged**